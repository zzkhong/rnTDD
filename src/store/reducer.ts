import {combineReducers} from 'redux';
import home from 'screen/Home/HomeReducer';

const reducers = combineReducers({
  home,
});

export type RootState = ReturnType<typeof reducers>;

export default reducers;
