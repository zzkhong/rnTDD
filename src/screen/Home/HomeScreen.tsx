import React, {useEffect, useState} from 'react';
import {SafeAreaView, Text, TouchableOpacity} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {RootState} from 'store/reducer';
import {setTitle} from './HomeActions';
import {Props} from './HomeTypes';

const HomeScreen = ({navigation}: Props) => {
  const [text, setText] = useState('');
  const dispatch = useDispatch();

  const title = useSelector((state: RootState) => state.home.title);

  useEffect(() => {
    setText('useEffect triggered');
  }, []);

  useEffect(() => {
    dispatch(setTitle('RNTDD'));
  }, [dispatch]);

  return (
    <SafeAreaView>
      <Text testID="homeScreen_title">{title}</Text>
      <Text testID="homeScreen_text">{text}</Text>
      <TouchableOpacity
        testID="homeScreen_detailButton"
        onPress={() => navigation.navigate('Details')}
      />
    </SafeAreaView>
  );
};

export default HomeScreen;
