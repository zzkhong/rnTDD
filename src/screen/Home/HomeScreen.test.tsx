import React from 'react';
import {act, create} from 'react-test-renderer';
import HomeScreen from './HomeScreen';

// necessary imports for redux
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import reducer from 'store/reducer';

// use fake timers to run async task call
// alternatively, we can create a file with jest.useFakeTimers()
// and modify the configuration in package.json OR jest.config.js
// by just adding "setupFiles": [ "<rootDir>/path_to_your_setup_file" ]
jest.useFakeTimers();

const store = createStore(reducer);
const navigation: any = {
  navigate: jest.fn(), // mock navigate function
};

const tree = create(
  <Provider store={store}>
    <HomeScreen navigation={navigation} />
  </Provider>,
);

test('#1 HomeScreen - Snapshot', () => {
  expect(tree).toMatchSnapshot();
});

jest.runAllTimers();

test('#2 useEffect call', () => {
  const text = tree.root.findByProps({testID: 'homeScreen_text'}).props;

  expect(text.children).toEqual('useEffect triggered');
  expect(text.children).not.toEqual('');
});

test('#3 Redux HomeState - setTitle', () => {
  expect(store.getState().home.title).toEqual('RNTDD');
});

test('#4 React Navigation - Navigate to Details', () => {
  const detailButton = tree.root.findByProps({
    testID: 'homeScreen_detailButton',
  }).props;
  act(() => detailButton.onPress());

  expect(navigation.navigate).toBeCalledWith('Details');
});
