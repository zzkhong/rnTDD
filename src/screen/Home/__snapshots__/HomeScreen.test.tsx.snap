// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#1 HomeScreen - Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text
    testID="homeScreen_title"
  >
    RNTDD
  </Text>
  <Text
    testID="homeScreen_text"
  >
    useEffect triggered
  </Text>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="homeScreen_detailButton"
  />
</RCTSafeAreaView>
`;
